### ISSUE TITLE
Algebraic Data Types (ADTs, Sealed Classes, Enum with associated values)

### ISSUE THREAD

**Issue by ZakTaccardi** [dart-lang/language#349]
ADTs are a very important part of any modern programming language.
 - Kotlin has [sealed classes](https://kotlinlang.org/docs/reference/sealed-classes.html#sealed-classes)
 - Swift has [enums with associated values](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/Enumerations.html)

Imagine you have a network call that submits login credentials. This call can:
 - log the user in
 - fail because the user entered incorrect wrong credentials
 - fail because the user doesn't have network connectivity
 - fail because of an unknown defect in your code or server broke API contract and the call failed

In Kotlin, the return value *could* be represented as the following `sealed class`

```kotlin
sealed class LoginResponse {
    data class Success(val authToken) : LoginResponse()
    object InvalidCredentials : LoginResponse()
    object NoNetwork : LoginResponse()
    data class UnexpectedException(val exception: Exception) : LoginResponse()
}
```

Kotlin's `when` statement unwraps this beautifully, [smart casting](https://kotlinlang.org/docs/reference/typecasts.html#smart-casts) `LoginResponse`.

```kotlin
fun on(loginResponse: LoginResponse) {
    return when (loginResponse) {
        is LoginResponse.Success -> {
            // direct access to auth token available w/o need to cast
            loginResponse.authToken
        }
        is LoginResponse.InvalidCredentials -> {

        }
        is LoginResponse.NoNetwork -> {

        }
        is LoginResponse.Unexpected -> {
            // direct access to exception available w/o need to cast
            loginResponse.exception
        }
    }
}
```

Sadly, languages like Java and Dart are unable to elegantly express this concept (ADTs). For a best case example, checkout out Retrofit's [`Response`](https://square.github.io/retrofit/2.x/retrofit/retrofit2/Response.html) object. It is a single pojo that contains both:

* The success [`.body()`](https://square.github.io/retrofit/2.x/retrofit/retrofit2/Response.html#body--)
* The [`.errorBody()`](https://square.github.io/retrofit/2.x/retrofit/retrofit2/Response.html#errorBody--)

Both these fields are nullable, but one must be non-null and the other must be null. To know which field can be accessed, you have to check the [`.isSuccessful()`](https://square.github.io/retrofit/2.x/retrofit/retrofit2/Response.html#isSuccessful--) flag. 

Now imagine the verbosity if this best case scenario has to scale to 4 or more possible results. A whole new enum class would have to be added to work around it...it's not pretty.

Please add support for ADTs to Dart. 

note: As an Android developer, I find that Flutter is an interesting proposition that is currently held back by the language.

**Comment by kevmoo** [comment]
CC @munificent

**Comment by tchupp** [comment]
@LinusU great example!

Dart has the idea of a refinement/type test/smart cast. I wonder if this could work similar to how Kotlin does this.

To re-use @LinusU's example:
```kotlin
when (_state) {
  is _ChooseProviderState -> ...
  is _LoadingState -> ...
  is _ChooseAccountsState -> ...
  is _ImportingState -> ...
}
```

Kotlin was originally designed on top of the JVM.
Java doesn't have the idea of smart casting, so this was a significant improvement:
```java
if (_state instanceof _ChooseProviderState) {
  _ChooseProviderState inner = (_ChooseProviderState) _state;
  ...
}
...etc.
```

Dart already has the idea of smart casting, so I have a feeling that implementing this should not be the most difficult. Agreeing on a syntax will be the fun part 😅 

(To steal what @LinusU said) Is there anything I could do to make this possible?

**Comment by renatoathaydes** [comment]
@LinusU @tchupp I agree that native support for ADTs would be nice and that it's much easier to work with ADTs in Rust/Kotlin/Swift/F# etc!

But at least the problem of not having compiler guarantees can be mostly worked around in Dart.

To continue on the example @LinusU posted above, you can add a method to the common type, `_State`, which lets you use it in a type-safe manner, ensuring you don't forget any case, by moving the `if/else` chain into a single place (a technique I use in Java normally, but works in Dart too - with the limitation that Dart doesn't have private constructors for public classes or final classes, which make this 100% safe in Java, but not in Dart, as you can still extend the base type elsewhere because of that if it's public non-final):

```dart
abstract class _State {
  T use<T>(
      T Function(_ChooseProviderState) useChooseProviderState,
      T Function(_LoadingState) useLoadingState,
      T Function(_ChooseAccountsState) useChooseAccountState,
      T Function(_ImportingState) useImportingState) {
    if (this is _ChooseProviderState) {
      return useChooseProviderState(this);
    }
    if (this is _LoadingState) {
      return useLoadingState(this);
    }
    if (this is _ChooseAccountsState) {
      return useChooseAccountState(this);
    }
    if (this is _ImportingState) {
      return useImportingState(this);
    }
    throw Exception('Invalid state');
  }
}
```

Now, your `build` method looks much nicer:

```dart
  @override
  Widget build(BuildContext context) {
    return _state.use(
        (chooseProvider) => _buildChooseProvider(context, chooseProvider),
        (loading) => _buildLoading(context, loading),
        (chooseAccounts) => _buildChooseAccounts(context, chooseAccounts),
        (importing) => _buildImporting(context, importing));
  }
```

[Dart Pad Full Example](https://dartpad.dartlang.org/8f477a2dfd48d70eeeace03023716638)

**Comment by ds84182** [comment]
Looks great! I'd suggest using `@required` named parameters though, so you don't have to depend on a particular argument order. It also makes it possible to remove subtypes in the future without getting... lots of unreadable errors from the analyzer.

**Comment by thosakwe** [comment]
Just adding OCaml's variant types to the conversation here.

A sum type in OCaml looks like:

```ocaml
type FeedWidgetState =
  | Loading
  | Failed of exn * stack_trace
  | Succeeded of tweet list

let build state context =
  match state with
    | Loading -> Flutter.circularProgressIndicator
    | Failed e _ ->
        let errMsg = Printexc.to_string e in
        Flutter.center (Flutter.text errMsg)
    | Succeeded accounts ->
         let render_account account =
           Flutter.listTile ~title: (Flutter.text account.handle)
         in
         Flutter.listView ~children: (List.map render_account accounts)
```

Obviously, the syntax is pretty different from Dart, but theoretically a translation could still look pretty Dart-y:

```dart
sum class FeedWidgetState {
   Loading,
   Failed(Object error),
   Succeeded(List<Account> accounts)
}

Widget build(BuildContext context) {
   // Or: match (await someComputation(2, "3")) as state
   return match state {
     Loading => CircularProgressIndicator(),
     Failed => Center(child: Text(state.error.toString())),
     Succeeded {
       return ListView(
         children: state.accounts.map(renderAccount).toList(),
       );
     }
   }
}
```

**Comment by nodinosaur** [comment]
`@sealed` annotation is now available as of [meta](https://pub.dev/packages/meta) `^1.1.7` 
(merged in on #11)

**Comment by nodinosaur** [comment]
How does this example look?
```
import 'package:meta/meta.dart';

@sealed
abstract class ViewState {
  ViewState reduce(ViewState state);

  T _when<T extends ViewState>(
    Object Function(IsIdle) isIdle,
    Object Function(IsLoading) isLoading,
    Object Function(Success) success,
    Object Function(Error) error,
  ){
    return <Type, Function>{
      IsIdle: isIdle,
      IsLoading: isLoading,
      Success: success,
      Error: error,
    }[runtimeType](this);
  }
}

class IsIdle extends ViewState {
  @override
  ViewState reduce(ViewState state) {
    return _when(
      (isIdle) => print('${isIdle.toString()}'),
      (isloading) => print('${isloading.toString()}'),
      (success) => print('$success.data'),
      (error) => print('{$error.msg}'),
    );
  }
}

class IsLoading extends ViewState {
  @override
  ViewState reduce(ViewState state) {
    return _when(
      (isIdle) => print('${isIdle.toString()}'),
      (isloading) => print('${isloading.toString()}'),
      (success) => print('$success.data'),
      (error) => print('{$error.msg}'),
    );
  }
}

class Success extends ViewState {
  Success({@required this.data});

  final Object data;

  @override
  ViewState reduce(ViewState state) {
    return _when(
      (isIdle) => print('${isIdle.toString()}'),
      (isloading) => print('${isloading.toString()}'),
      (success) => print('$success.data'),
      (error) => print('{$error.msg}'),
    );
  }
}

class Error extends ViewState {
  Error({@required this.message});

  final String message;

  @override
  ViewState reduce(ViewState state) {
    return _when(
      (isIdle) => print('${isIdle.toString()}'),
      (isloading) => print('${isloading.toString()}'),
      (success) => print('$success.data'),
      (error) => print('{$error.msg}'),
    );
  }
}
```

**Comment by andrewackerman** [comment]
ADTs as a whole can mostly be represented in dart currently using inheritance. The only thing that's limiting those approaches' success is their verbosity. However, I think this would largely be alleviated by the addition of data classes and the addition of a `when` construct which can be an extension of a `switch` block.

Take the Kotlin code in the OP, for example. If Dart had data classes and a `when` construct, the Dart equivalent of the Kotlin code might look something like this:

    abstract data class LoginResponse;
    data class Success(String authToken) extends LoginResponse;
    data class InvalidCredentials extends LoginResponse;
    data class NoNetwork extends LoginResponse;
    data class UnexpectedException(Error error) extends LoginResponse;

    // ...

    on(LoginResponse loginResponse) {
      when (loginResponse) {
        is Success:
          // direct access to auth token available w/o need to cast
          return loginResponse.authToken;
        is InvalidCredentials:
          return null;
        is NoNetwork:
          return null;
        is Unexpected:
          return loginResponse.error;
      }
    }

If Dart gets support for pattern matching in switch blocks, this might not even need a dedicated `when` construct as the type matching can be made in the case statements.

**Comment by nodinosaur** [comment]
Yes, I agree that we are missing `when` in the example I have, the `_when` method is not exhaustive, nor does it catch any additional sub-types added to the parent class with the `@sealed` annotation.
For example, adding an isPaused does not give any IDE or Compiler hints:
```
class IsPaused extends ViewState {
  @override
  ViewState reduce(ViewState state) {
    return _when(
      (isIdle) => print('isIdle'),
      (isloading) => print('isloading'),
      (success) => print('${success.data}'),
      (error) => print('${error.message}'),
    );
  }
}
```

What I would like to see if there is an official _Dart_ example of how we ought to be implementing `@sealed`? (cc. @srawlins)

**Comment by andrewackerman** [comment]
`@sealed` just means the class isn't supposed to be extensible, i.e. it can't be the superclass to any other class. The ADT behavior of sealed classes in Kotlin just takes advantage of the fact that sealed classes in Kotlin can have child classes declared, but they must be in the same file as the sealed class. (So, in reality, Kotlin is also using inheritance to achieve ADT-like behavior. There are just a couple more features Kotlin offers to make these "ADTs" more tidy to use overall like exhaustive type checking.)

**Comment by xsahil03x** [comment]
I did something similar to what @avilladsen and @renatoathaydes mentioned above and was satisfied with the results. To avoid the boilerplate, I built a [code generator](https://github.com/xsahil03x/super_enum) with my colleague which generates these classes by annotating Enums.

```dart
@superEnum
enum _MoviesResponse {
  @Data(fields: [DataField('movies', Movies)])
  Success,
  @object
  Unauthorized,
  @object
  NoNetwork,
  @Data(fields: [DataField('exception', Exception)])
  UnexpectedException
}
```
where:-

`@Data()` marks an enum value to be treated as a Data class.

- One should supply a list of possible fields inside the annotation.

- If you don't want to add fields, use `@object` annotation instead.

- Fields are supplied in the form of `DataField` objects.

- Each `DataField` must contain the `name` and the `type` of the field.

- If the field type needs to be generic use `Generic` type and annotate the enum value with `@generic` annotation.

`@object` marks an enum value to be treated as an object.

**Comment by renatoathaydes** [comment]
Very cool @xsahil03x ... the existing solutions to this are probably not as nice.

https://github.com/google/built_value.dart/blob/master/example/lib/enums.dart

https://github.com/werediver/sum_types.dart

Would be nice if one alternative became the standard though.

**Comment by tvh** [comment]
Another alternative is https://github.com/factisresearch/sum_data_types. It offers both sum- and product-types. Unfortulately the Syntax for the constructors is a bit cumbersome.

**Comment by renatoathaydes** [comment]
Everyone wants to solve this problem :D here's a few more existing libraries from https://github.com/dart-lang/language/issues/546#issuecomment-525382550

> If existing packages are interesting to see, here's a bunch of them:
> 
>     * [sealed_unions](https://pub.dev/packages/sealed_unions)
> 
>     * [tuple](https://pub.dev/packages/tuple)
> 
>     * [dartz](https://pub.dev/packages/dartz)

**Comment by nodinosaur** [comment]
Seems to me that the effort creating a new library might have been better spent improving [Sealed Unions](https://pub.dev/packages/sealed_unions) as it is already being widely used.
Still at least there is choice :)

**Comment by satvikpendem** [comment]
[Remi Rousselet](https://github.com/rrousselGit), who made the great [provider](https://pub.dev/packages/provider), [functional_widget](https://pub.dev/packages/functional_widget), and [flutter_hooks](https://pub.dev/packages/flutter_hooks) libraries, among others, recently launched the [freezed](https://pub.dev/packages/freezed) package that includes immutable classes and also algebraic data types, along with an exhaustive `when` function to switch between the unions.

You can create immutable classes using an abstract class with the `@immutable` annotation and the specified properties:

```dart
@immutable
abstract class Person with _$Person {
  factory Person(String name, int age) = _Person;
}
```

For ADTs, create an abstract class with factories specifying their values, and use `when` to pattern match them:

```dart
@immutable
abstract class Model with _$Model {
  factory Model.first(String a) = First;
  factory Model.second(int b, bool c) = Second;
}

var model = Model.first('42');

print(
  model.when(
    first: (String a) => 'first $a',
    second: (int b, bool c) => 'second $b $c'
  ),
); // first 42
```

There's also `maybeWhen`, similar to other languages where there is a default case, using the `orElse` keyword:

```dart
@immutable
abstract class Union with _$Union {
  const factory Union(int value) = Data;
  const factory Union.loading() = Loading;
  const factory Union.error([String message]) = ErrorDetails;
}

var union = Union(42);

print(
  union.maybeWhen(
    null, // ignore the default case
    loading: () => 'loading',
    // did not specify an `error` callback
    orElse: () => 'fallback',
  ),
); // fallback
```

It looks to be a pretty cool library. I'm glad that Dart has code generation so that we can create language features via libraries instead of just waiting for them by the official Dart team, although language features would be preferred over libraries of course, if only to cut down the amount of generated code and boilerplate.

**Comment by flukejones** [comment]
Chiming in to say that having spent the last few years neck deep in Rust and Swift, coming to Dart without ADT feels a little like having a hand tied behind your back.

This really is a must have language feature and would enable easier creation of robust code.

**Comment by nodinosaur** [comment]
True, but as [Brian](https://github.com/dart-lang/language/issues/349#issuecomment-400959536) mentioned we have had [sealed_unions](https://pub.dev/packages/sealed_unions) for nearly 2 years now and other have managed to use them successfully during that time.
However, I am glad to see that the community at large is helping the language.

**Comment by flukejones** [comment]
That may be the case but I'm sure a language level solution would be much better than ad-hoc with third-party libraries, and would also expose opportunities for compiler level optimizations.

**Comment by nodinosaur** [comment]
Yes, ultimately that is what we all want & hopefully we won't be waiting too much longer :)

**Comment by pedromassango** [comment]
Where this is being worked on or tracked?

**Comment by kevmoo** [comment]
> Where this is being worked on or tracked?

We are VERY much heads-down on null-safety at the moment. If you're subscribed to this issue, you'll see updates as they happen.

**Comment by pedromassango** [comment]
> > Where this is being worked on or tracked?
> 
> We are VERY much heads-down on null-safety at the moment. If you're subscribed to this issue, you'll see updates as they happen.

Thank you. I really want to stay up to date with this.

**Comment by nguyenxndaidev** [comment]
I like Flutter, but Dart language in general, especially Enum in Dart, is holding me back :(. Hope that, Dart will soon be comparable with modern languages like Kotlin, Swift or C#.

**Comment by jodinathan** [comment]
well, I guess technically Dart has some of this implemented on try catch clause.

I guess what we can do now is use NNBD and feedback them so it can go stable soon.

**Comment by veggiedefender** [comment]
hiya, any updates on ADT support? :slightly_smiling_face:

**Comment by pedromassango** [comment]
> hiya, any updates on ADT support? :slightly_smiling_face: 

The Dart team is focused in NNBD now. All we can now is to thumbs up this issue and keep an eye in here.

**Comment by Producer86** [comment]
for the time being, a visitor can also be used to get compile-time errors about missing cases.
```dart
abstract class ResponseVisitor<R> {
  R visitSuccess(SuccessResponse response);
  R visitFailure(FailureResponse response);
}

abstract class Response {
  R accept<R>(ResponseVisitor<R> visitor);
}

class SuccessResponse implements Response {
  SuccessResponse(this.data);

  final int data;

  @override
  R accept<R>(ResponseVisitor<R> visitor) {
    return visitor.visitSuccess(this);
  }
}

class FailureResponse implements Response {
  FailureResponse(this.message);

  final String message;

  @override
  R accept<R>(ResponseVisitor<R> visitor) {
    return visitor.visitFailure(this);
  }
}

class ServiceClass {
  Response getData() {
    try {
      return SuccessResponse(42);
    } on Exception {
      return FailureResponse('No data found');
    }
  }
}

class WorkerClass implements ResponseVisitor<void> {
  WorkerClass(this.service);

  final ServiceClass service;

  @override
  void visitFailure(FailureResponse response) {
    print('Error happened: ${response.message}');
  }

  @override
  void visitSuccess(SuccessResponse response) {
    print('Result arrived: ${response.data}');
  }

  void doWork() {
    service.getData().accept(this);
  }
}

void main() {
  final service = ServiceClass();
  WorkerClass(service).doWork();
}
```

**Comment by munificent** [comment]
> The Dart team is focused in NNBD now. All we can now is to thumbs up this issue and keep an eye in here.

Yes, and we do very much appreciate your patience. :) We can't wait to finish up NNBD and move onto other language features.

**Comment by maks** [comment]
Thanks @munificent I'm certainly very much appreciate all the efforts the Dart team is putting in for NNBD! And NNBD already makes the "enums+extensions with `switch`'s" pattern so much more useful for the kind of things that I'm used to using sealed classes.

And in the meantime I spotted some new lints in Dart 2.9 that I think might help with some of the uses described here until there is built in language support: [exhaustive_cases](https://github.com/dart-lang/language/issues/349#issuecomment-673124362) and [no_default_cases](https://dart-lang.github.io/linter/lints/no_default_cases.html).

**Comment by pedromassango** [comment]
Related to https://github.com/dart-lang/language/issues/83

**Comment by bubnenkoff** [comment]
Does Algebraic DT include monads like Either?

**Comment by minikin** [comment]
@bubnenkoff check this pub https://pub.dev/packages/result_type it might help.

**Comment by nodinosaur** [comment]
I tend to use https://pub.dev/packages/dartz for functional programming types. 

Functional programming in Dart

> Type class hierarchy in the  spirit of cats, scalaz and the standard Haskell libraries
> • Immutable, persistent collections, including IVector, IList, IMap, IHashMap, ISet and AVLTree
> • Option, Either, State, Tuple, Free, Lens and other tools for programming in a functional style
> • Evaluation, a Reader+Writer+State+Either+Future swiss army knife monad
> • Type class instances (Monoids, Traversable Functors, Monads and so on) for included types, as well as for several standard Dart types
> • Conveyor, an implementation of pure functional streaming

**Comment by bubnenkoff** [comment]
thanks! But is there any plan to add something similar to Dart to get ADT
out of the box?

чт, 4 февр. 2021 г. в 12:50, George.M <notifications@github.com>:

**Comment by flukejones** [comment]
All of these external packages feel like bandaid solutions to something that would likely be able to optimize heavily if it were in the language as a standard feature.

**Comment by adrianboyko** [comment]
Sum types are not yet in the [language funnel](https://github.com/dart-lang/language/projects/1)  :(

**Comment by Abion47** [comment]
@adrianboyko Check again under "Patterns and other features".

**Comment by Nikitae57** [comment]
Another vote for this feature

**Comment by ilexbor-codex** [comment]
How long can I wait? To write secure code, you have to suffer. Sealed classes have been implemented in all modern languages for a long time.

**Comment by jjoelson** [comment]
I'm not sure if commenting here will be helpful all, but I'm a big +1 to this feature request. When it comes to modeling a problem domain in a modern programming language, I'd say ADTs are very nearly as important as null safety. With null-safety implemented, I can't imagine any language feature that ought to be a higher priority than ADTs.

Flutter being such an important use of Dart implicitly puts Dart in direct competition with Swift and Kotlin. Both of those languages not only support ADTs, but have a strong developer culture of using them heavily. That means that while there is a lot about the developer experience that improves when one switches from native to Flutter, the lack of ADTs in Dart is a glaring regression.

**Comment by hepin1989** [comment]
Scart

**Comment by hepin1989** [comment]
I prefer we take a look at the Scala 3:
https://docs.scala-lang.org/tour/case-classes.html
https://docs.scala-lang.org/tour/pattern-matching.html
https://docs.scala-lang.org/overviews/core/value-classes.html#inner-main
https://docs.scala-lang.org/scala3/book/types-union.html
https://docs.scala-lang.org/scala3/book/types-intersection.html
https://docs.scala-lang.org/scala3/reference/enums/enums.html
https://docs.scala-lang.org/scala3/book/types-opaque-types.html#inner-main

**Comment by Abion47** [comment]
@hepin1989 

* Case classes: 
  * #314
* Pattern matching: 
  * #546 
* Value classes:
  * In a way, Dart already has this with `const` constructors.
* Types union:
  * This issue, #83 
* Types intersection:
  * This issue and maybe #83 as well, or maybe it's out of scope there.
* Enums:
  * A lot of issues tracking enum improvements, did you have something specific in mind?
* Opaque types:
  * This can be done today with typedefs and extension methods.

**Comment by JarvanMo** [comment]
This is really awesome for Kotlin developers. You can say  the current built-in types are enough for us but *ENOUGH* is far less than we need. As a newer language, I believe we can do better in grammar sugar. I miss Kotlin very single time I write Flutter! I want data classes and removing the `;`.

**Comment by munificent** [comment]
> I want data classes and removing the `;`.

These are tracked on separate issues.

> And also enum class bring nothing but string and index!

The next stable release of Dart will [allow you to define fields, methods, and other members on enum types](https://github.com/dart-lang/language/blob/master/accepted/future-releases/enhanced-enums/feature-specification.md). It's out in the beta channel now.

**Comment by marcguilera** [comment]
I think the overall experience writing dart is good but I do miss some things and by reading here it looks like I'm not alone. It would be great if we could have the equivalent of:
- Sealed interfaces.
- Type matching in switch (or some other keyword like `when`) which could be exhaustive for sealed interfaces.
- Data classes or records to represent data with automated copyWith, equals and hashCode. I think the main use case is for immutable data, but maybe it doesn't need to be restricted to it.

Packages like `freezed` are great but having to rely on code generation for this common cases feels sub optimal in my opinion. 

The good news is that it looks like there are issues for most of these and are in various levels of discussions already. It's great we are getting fields in enums in the next release!

**Comment by renatoathaydes** [comment]
> It's great we are getting fields in enums in the next release!

That was released in Dart 2.17 a few days ago. See the announcement: https://medium.com/dartlang/dart-2-17-b216bfc80c5d

**Comment by munificent** [comment]
Closing this because records and patterns are now enabled on the main branch of the Dart SDK! 🎉
